AKS provisioning extension

mantis-toboggan-md opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #8965 

Pre-existing AKS bug fixes/feature requests also covered by this move:
#6071 
#7913 
#9416 
#9433 
#9526 
#9525 
#8713 
#3412 
#8750 
#6879 
#10125 

### Occurred changes and/or fixed issues
This PR re-creates the aks cluster provisioning flow in vue. Functionality should largely be the same.

### Technical summary
This PR adds AKS prov as an extension. We already have some RKE2 extension points, but this is a slightly different use case, where we want to replace the _entire_ provisioning form. I've added a mechanism for an extension provisioner to provide its own form component, because this seemed the most straightforward way of keeping routing to/from this aks form consistent with the rest of cluster provisioning. 

### Areas or cases that should be tested
AKS Provisioning generally...
Unlike in ember the credential and region can be changed without re-doing the whole form so it's probably a good idea to play around with that and verify that the ux generally makes sense.

### Areas which could experience regressions
The logic used to determine when an ember page is shown for various cluster types was touched, so verifying that other cluster types still load the ember creation form when expected would be good

### Known issues
* The vm size dropdown loads terribly. This is a bug with LabeledSelect, being tracked in #9668 
* The `upstreamDiff`brought over from ember has a known bug around deleting items from arrays: #6879 this will be covered in a subsequent PR  
* This code needs tests! I think I've written the main AKS form component in such a way that tests can be written without supplying actual AKS creds. I'm working on them today/next week, but I wanted to put up the PR anyway because it's going to take time to review. But if reviewers want to wait for me to add tests, my feelings wont be hurt.

### Screenshot/Video
https://github.com/rancher/dashboard/assets/42977925/76d35511-5ac9-42c3-a79c-78a33c4d20d4
![Screen Shot 2023-09-08 at 9 35 09 AM](https://github.com/rancher/dashboard/assets/42977925/5ad3ed42-0f86-4082-bc35-d90bf5015d45)
![Screen Shot 2023-09-08 at 9 34 56 AM](https://github.com/rancher/dashboard/assets/42977925/a03da0ce-1df0-41a1-8d28-e20cda276e4f)

mantis-toboggan-md commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1038 @@
+import Accordion from '@pkg/aks/components/Accordion.vue';
+import AksNodePool from '@pkg/aks/components/AksNodePool.vue';
+
+const defaultNodePool = {
```
These defaults are copied over from Ember

mantis-toboggan-md commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1038 @@
+      this.$set(this.config, 'clusterName', name);
+    },
+
+    onMembershipUpdate(update: any) {
+      this.$set(this, 'membershipUpdate', update);
+    },
+
+    async saveRoleBindings() {
+      if (this.membershipUpdate.save) {
+        await this.membershipUpdate.save(this.normanCluster.id);
+      }
+    },
```
The rolebinding save flow is pretty much copied from rke2.vue - in the future we should consider refactoring ClusterMembershipEditor.vue to make it easier to implement in other cluster provisioning extensions but that seems outside the scope of this PR

richard-cox reviewed (changes requested):
Haven't dived in to much code yet, but wanted to give a review to keep things ticking

- ~rke2 world asks the user to create or select credentials before progressing to the next screen. the new flow shows this fake step~
  - ~there's some weirdness when the user tries to create credentials (double footers) when on the page~
  - ~haven't tried this, but what happens when there's no credentials to start with?~
- things missing from old UI
  - the old ui splits Cluster Location (now Region) into two buckets - HA and Other. if the user selects one from `Other` ...
    - ~they get a warning `No availablity zones are available for the selected region. For the most resiliency it is recommended to select a High Availablity region.` (just noticed the spelling mistakes)~
    - the node pools `Availability Zones` are disabled
      - a red warning is shown, but the AV component can be interacted with 
  - ~cluster resource group~ and node resource group had some really useful tooltips
    - node resource group message is missing items in angle brackets (shows as MC___ instead of `MC_<resource-group1>...`)
  - ~Network Policy had a tooltip `Azure Network Policy is only available when Azure CNI is selected as the Network Plugin`~
    - ~the azure network policy is disabled if network plugin is not azure cni~
    - ~setting network policy and then network plugin --> then changing network plugin from azure cni resets network policy to none. there might be some autoselecting of the checkboxes below as well~
  - Network Plugin of `Kubenet` should hide (and clear?) virtual network, kube service address range, kube pod address ranve, kube dns service address and docker bidge address inputs
  - ~Project Network Isolation should be disabled if there's no policy (there's a tooltip as well)~
  - ~checking Enabled Private Cluster shows a warning `AKS private clusters is an advanced use case which requires to be running on the same NAT as the AKS nodes. Before enabling private clusters ensure you have read through the <html link>AKS cluster provisioning</link> docs for instructions on configuring AKS so nodes may communicate with .`~
  - Node Pools
    - `Mode` - `Primary Pool - System` has been dropped. Was this ever editabled?
    - ~Taint key value pair in old ui had a small message `Taint must be unique by key and effect pair.`. Looks like it doesn't validate this, just warns~
  - ~Comment Only. Each expandable section had a brief intro to content. Think all were superfluous though~
  - There was an optional `Description` input field in the old ui
- The location of the `Add Cluster Member` modal when `Add`ing a `Cluster Members` shows up half way down the screen
  - refresh on the page, scroll down to the Cluster Members, expand, click Add.
  - close modal, scroll around, clidk Add, modal is in the right place
- UX Wise
  - ~Could we show the expandable sections in tabs (though this could be phase 2). If not could there be a smidge more vertical spacing?~
- Node Pools
  - ~The input fields for `Enable Auto Scaling` input fields appear underneath the others, and feel kinda disjointed. Can these all be on a single new row?~
  - ~Adding a Taint and changing the Effect will cause the input field widths to jiggle about a bit~
  - Taints expandable section had a useful description that's been lost `Taints mark that the node should not accept any pods that do not tolerate the taints.`. Grammar wise it's not quite right, but would be good to have similar/same in the new UI.
    - There's a placeholder there now, but that's used for `Tain must be unique by key and effect pair`
  - Similar to above, Labels has `Configure labels for all nodes in this Node Pool`. 50:50 on this. might be useful to remind user they apply to all nodes, rather than it's labels for the node pool resource itself?
  
  
Tracking things I need to test / progress
- Create flow as non-admin (what is shown for the initial cluster member?)
- Create flow when no credentials found
- Validate existing issues fixed by this
- Edit Cluster
  - edit name?
- Upgrade Cluster
- Deploying a cluster in a Region in the `Other` section

richard-cox reviewed (changes requested):
Ok, finished going through code. There's still some testing i'd like to do (at bottom) but will wait for code changes first.

Old UI
- Edit Cluster
  - ~the old flow let the user select the credentials again. not sure if these where then associated with the cluster, were just used in that context or were ignored because it used the same flow as create~
  - `Member Roles` had a warning `Note: Removing User will not remove their project permissions`
    - I think this is trying to say removing a cluster role will not mean users lose project/namespace roles (and therefor can still see cluster)?
    - Update: Ok, lets track that in another issue. It's an important warning that we should probably be showing in other cases as well
  - ~Max Pods Per Node isn't editable~
    
Bugs
- ~Edit Cluster~
  - ~Cluster with name `rc-aks-2023-09-11` falsely gives the error `Name must start and end with an alphanumeric`~
- ~Values added to to create cluster page are persisted over visits~
  - ~Go to the create cluster page, enter some values for a number of fields, click on side nav clusters --> back to create cluster page. The same values previously entered are shown.~
  - ~Mostly resolved, but i saw `AV Zone` is still kept over nav away --> back.~
  - ~To Test~
    - ~Think this might be something to do with the way vue reuses the `component`? Might be fixed with a `:key`~
  
Tracking things I need to test / progress / pass to QA
- Create flow as non-admin (what is shown for the initial cluster member?)
- Create flow when no credentials found
- Validate existing issues fixed by this
- Edit Cluster
  - edit name?
  - Create cluster as non-admin, can (only) user/member be removed?
- Upgrade Cluster
- Note for QA - someone with a lot of AKS experience should ensure no regressions
- Test other hosted providers
- Validate kube version flow
- Member management
- Errors that cause API failure are shown correctly
- Deploying a cluster in a Region in the Other section

mantis-toboggan-md commented on pkg/aks/components/Taint.vue:
```diff
@@ -0,0 +1,104 @@
+<script lang="ts">
```
I think that would get confusing: in the Rancher world node taints are an object with `key`, `value`, and `effect` properties, while here they're a string in the format key=value:effect

mantis-toboggan-md commented on pkg/aks/package.json:
```diff
@@ -0,0 +1,24 @@
+    "last 2 versions",
+    "not dead"
+  ]
+}
```
Yeah that seems to be what Elemental is doing. I've updated this to use the aks icon.

richard-cox replied:
Looks like it also needs config pkg/aks/index.ts

```
  // Built-in icon
  plugin.metadata.icon = require('./icon.svg');
```

That gives the following 

![image](https://github.com/rancher/dashboard/assets/18697775/3fc7d6a7-12fd-4d12-a6d8-e970d7219763)

mantis-toboggan-md commented on shell/edit/provisioning.cattle.io.cluster/SelectCredential.vue:
```diff
@@ -206,6 +214,10 @@ export default {
+    },
+
+    defaultOrCancel() {
+      this.defaultOnCancel ? this.credentialId = this.filteredCredentials[0]?.id : this.cancel();
```
Honestly idk what I was on with this. I've removed this change and updated the aks credential implementation to match rke2

mantis-toboggan-md commented on shell/edit/provisioning.cattle.io.cluster/index.vue:
```diff
@@ -205,12 +212,14 @@ export default {
         }
         if ( this.subType ) {
+          // if driver type has a custom form component, don't load an ember page
+          if (this.selectedSubType.component) {
```
Nice catch/sorry you had to debug that...I've updated the logic around subtypes a bit so that if an extension provisioner doesn't have the rke2 'group' set its visibility isn't dependent on the rke1/rke2 toggle. I think I've done this in a way that the default behaviour is unchanged/we wont break some pre-existing rke2-only extension provisioner.

mantis-toboggan-md commented on pkg/aks/util/validators.ts:
```diff
@@ -0,0 +1,118 @@
+  };
+};
+
+// cluster name
```
Fixed the validator + added unit tests for it

mantis-toboggan-md commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1050 @@
+
+    // filter out versions outside ui-k8s-supported-versions-range global setting and versions < current version
+    // sort versions, descending
+    aksVersionOptions() {
```
I'm not sure off the top of my head exactly which bugs you're referencing but I added unit tests around the version setting, sorting, and prevention of downgrades.

mantis-toboggan-md commented on pkg/aks/components/AksNodePool.vue:
```diff
@@ -0,0 +1,317 @@
+          v-model="pool.osType"
+          :options="[]"
+          label-key="aks.nodePools.osType.label"
+          mode="view"
```
Thanks for reminding me, this is always disabled in Ember and I forgot to go back and investigate why...From what I can tell going off git history, (sparse) github comments, and QA test plan confluence history, AKS windows support was initially planned then scrapped. This input was not editable in any Rancher release. 

I've removed the input entirely. I'm not sure if we need to explicitly stated that the nodes created here are going to be Linux nodes? My understanding is that windows k8s nodes are somewhat...niche and less supported. But I could be convinced otherwise.

richard-cox commented on pkg/aks/index.ts:
```diff
@@ -0,0 +1,15 @@
+import { AKSProvisioner } from './provisioner';
+
+// Init the package
+export default function(plugin: IPlugin): void {
```
RE Icon shown in the dashboard's extension list, looks like it also needs config here to show

```
  // Built-in icon
  plugin.metadata.icon = require('./icon.svg');
```

That gives the following 

![image](https://github.com/rancher/dashboard/assets/18697775/3fc7d6a7-12fd-4d12-a6d8-e970d7219763)

richard-cox commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1197 @@
+    }
+  },
+
+  // AKS provisioning needs to use the norman API - a provisioning cluster resource will be created byt he BE when the norman cluster is made but v2 prov clusters don't contain the relevant aks configuration fields
```
small typo `byt he`

richard-cox commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1197 @@
+
+const _NONE = 'none';
+
+// export default (Vue as VueConstructor<
```
Comment can be removed

richard-cox commented on pkg/aks/components/CruAks.vue:
```diff
@@ -0,0 +1,1197 @@
+          return systemPool ? undefined : this.t('aks.nodePools.mode.systemRequired');
+        },
+
+        availabilityZoneSupport: () => {
```
a few bits of commented out code

richard-cox reviewed (changes requested):
I've gone through the previous two review comments and crossed out resolved items. All comments have also been updated and resolved if could.

I hit the following things as well

- After entering junk credentials and clicking create there's a warning at the top... but the screen changes to the create page still
- It's possible to create creds with no name. Do we care about this?
- In the old ui when Network Plugin `Azure CNI` is selected and the 5 additional input fields are shown, 3 of them start off with values. They match the examples, should they be defaults in the new UI?
- Create Cluster with a cluster member --> there's one request to `/v3/clusterroletemplatebindings` which fails with missing `clusterId` param. there's no request to create the cluster.
  - I then removed the added role (just leaving admin) and clicked Create. The user was returned to the cluster list, there was still no request to create the cluster
    - Update - this happens even when no roles were changed

mantis-toboggan-md commented:
>   * there's some weirdness when the user tries to create credentials (double footers) when on the page
>   * haven't tried this, but what happens when there's no credentials to start with?

I believe I've fixed any wonkiness with credentials: the flow should match rke2 behavior now.


   
>   * the old ui splits Cluster Location (now Region) into two buckets - HA and Other. if the user selects one from `Other` ...
>     
>     * they get a warning `No availablity zones are available for the selected region. For the most resiliency it is recommended to select a High Availablity region.` (just noticed the spelling mistakes)
>     * the node pools `Availability Zones` are disabled

Nice catch, I completely missed this: on my setup the region dropdown auto-scrolls right past the first header. Sadly, this list has to be hard-coded. I've brought in an updated list from Azure docs. I added some validation around this as well. The default node pool spec includes availabilityZones; I've changed it to not do so when adding new node pools if the selected region doesn't support them.

https://github.com/rancher/dashboard/assets/42977925/6a6bcc26-5b08-497f-b26b-e9ec628a1c6e


>   * cluster resource group and node resource group had some really useful tooltips

![Screen Shot 2023-11-21 at 2 49 39 PM](https://github.com/rancher/dashboard/assets/42977925/07c5d595-0e60-4f40-92ef-aaa96857cc28)
![Screen Shot 2023-11-21 at 2 49 33 PM](https://github.com/rancher/dashboard/assets/42977925/2d2cc893-ec39-443d-85d6-d87e2b99f196)


>   * Network Policy had a tooltip `Azure Network Policy is only available when Azure CNI is selected as the Network Plugin`
>     
>     * the azure network policy is disabled if network plugin is not azure cni
>     * setting network policy and then network plugin --> then changing network plugin from azure cni resets network policy to none. there might be some autoselecting of the checkboxes below as well
>   * Project Network Isolation should be disabled if there's no policy (there's a tooltip as well)

https://github.com/rancher/dashboard/assets/42977925/3f83786d-c713-45fb-ace3-24494ecc67ec


>   * checking Enabled Private Cluster shows a warning `AKS private clusters is an advanced use case which requires to be running on the same NAT as the AKS nodes. Before enabling private clusters ensure you have read through the <html link>AKS cluster provisioning</link> docs for instructions on configuring AKS so nodes may communicate with .`

![Screen Shot 2023-11-21 at 2 57 49 PM](https://github.com/rancher/dashboard/assets/42977925/3d64528f-db60-4bf5-a6aa-d1cbcb5ce832)
(my custom private label right now is 'abc')

>   * Node Pools
>     
>     * `Mode` - `Primary Pool - System` has been dropped. Was this ever editabled?

Previously it was impossible to remove the first node pool, which seemed to be an arbitrary restriction set by the UI and requested gone, https://github.com/rancher/dashboard/issues/6071 . Now you can remove the first node pool, but you do need at least one 'system' pool so I added validation:


https://github.com/rancher/dashboard/assets/42977925/6fca1a9d-393e-4ae4-a06e-359148557b99



>     * Taint key value pair in old ui had a small message `Taint must be unique by key and effect pair.`. Looks like it doesn't validate this, just warns

![Screen Shot 2023-11-21 at 3 03 13 PM](https://github.com/rancher/dashboard/assets/42977925/9ee70fa8-0f59-477a-91a2-af3ca821f0d7)

>   * Comment Only. Each expandable section had a brief intro to content. Think all were superfluous though

I discussed this with Ken when we were talking about the accordions more generally. Those subheaders don't include useful information the overwhelming majority of the time so we've opted to leave them out entirely.

>   * There was an optional `Description` input field in the old ui
> * The location of the `Add Cluster Member` modal when `Add`ing a `Cluster Members` shows up half way down the screen

Could this go into a separate issue? That component is also used in rke2 and I see the same behavior from it there.

> * UX Wise
>   
>   * Could we show the expandable sections in tabs (though this could be phase 2). If not could there be a smidge more vertical spacing?

Ken is generally opposed to increased tabification - I've added more spacing around the accordions.

> * Node Pools
>   
>   * The input fields for `Enable Auto Scaling` input fields appear underneath the others, and feel kinda disjointed. Can these all be on a single new row?

Do you mean something like this?
https://github.com/rancher/dashboard/assets/42977925/c780e390-d259-4b99-b7e7-d46dc164e7c1


>   * Adding a Taint and changing the Effect will cause the input field widths to jiggle about a bit

https://github.com/rancher/dashboard/assets/42977925/2bc1b121-69d3-45be-a669-814532795807



>   * Taints expandable section had a useful description that's been lost `Taints mark that the node should not accept any pods that do not tolerate the taints.`. Grammar wise it's not quite right, but would be good to have similar/same in the new UI.
>   * Similar to above, Labels has `Configure labels for all nodes in this Node Pool`. 50:50 on this. might be useful to remind user they apply to all nodes, rather than it's labels for the node pool resource itself?

I thought both of these were kinda just there to have _something_ in the accordion subheader. There's a lot of that going on in the UI. I left them out as we're not using them in RKE2 and a taint is a standard kubernetes concept/easily googled

mantis-toboggan-md commented:
>Network Plugin of Kubenet should hide (and clear?) virtual network, kube service address range, kube pod address ranve, kube dns service address and docker bidge address inputs

There was a request to expand kubenet functionality: https://github.com/rancher/dashboard/issues/8750

mantis-toboggan-md commented:
>the node pools Availability Zones are disabled
>a red warning is shown, but the AV component can be interacted with

This input should become disabled if the user clears out any selected availability zones. I didn't want the availability zone selections to be automatically cleared upon selecting a region without support, particularly since they may be configured on a nodepool that isn't currently being displayed. I also didn't want to force the user to switch back to an AZ-supportive region in order to clear out their AZ selections.

>Edit Cluster
>the old flow let the user select the credentials again. not sure if these where then associated with the cluster, were just >used in that context or were ignored because it used the same flow as create

Interestingly, the credential _is_ used in the aks configuration, but it seems you can change it on an existing cluster. 

>It's possible to create creds with no name. Do we care about this?
I don't think so; this was done intentionally (`metadata.generateName` is set in the cloud cred component). It's consistent with the UI as well.

>In the old ui when Network Plugin Azure CNI is selected and the 5 additional input fields are shown, 3 of them start off with values. They match the examples, should they be defaults in the new UI?

I checked how this works when creating an AKS cluster directly in the azure portal, and these defaults are also set there. Since we're now allowing these fields to be set when kubenet is selected, I verified that the cluster seems to provision fine with these default values + kubenet plugin.

>Member Roles had a warning Note: Removing User will not remove their project permissions
I think this is trying to say removing a cluster role will not mean users lose project/namespace roles (and therefor can still see cluster)?

This component is used across provisioning so I'd rather track any changes specific to it in another issue if you're amenable

>Edit Cluster
edit name?

I thought it was weird but this works in the old UI so I included it, didn't want less functionality here than in the old UI. `<norman cluster>.aksConfig.clusterName` wont change, but `<norman cluster>.name` can be edited.

>Mostly resolved, but i saw AV Zone is still kept over nav away --> back.
To Test
Think this might be something to do with the way vue reuses the component? Might be fixed with a :key

This seemed to be related to how I was setting default values on the norman cluster. Should be good now

>Create Cluster with a cluster member --> there's one request to /v3/clusterroletemplatebindings which fails with missing clusterId param. there's no request to create the cluster.

This worked and then I broke it while refactoring slightly to get rid of a linting error! Good job you retested.

mantis-toboggan-md commented:
This PR should also cover a newly-filed issue: https://github.com/rancher/dashboard/issues/10125

richard-cox reviewed (changes requested):
I've got through review comments and inline comments, only one left from above (to create an issue for a lost notice)

- Tried creating a cluster, UK South, AV Zone 1, multiple roles assigned to a non-admin member. The cluster create succeeds but the 3 additional cluster role bindings fail with 404s
  - ![image](https://github.com/rancher/dashboard/assets/18697775/37ef56fd-0c69-491d-b7fb-25fe8bf4cff9)
  - Editing the cluster 
- Not sure if this was a blip, but i edited a fresh cluster to increase the sole node pool node count from 1 to 2 and it wasn't shown in the PUT aksConfig.nodePools[0].count
  - the create and edit was in the same session, no refresh
  - editing the cluster a second time, changing count from 1 to 2, worked fine
- Possibly related to above, setting the node pool node count to 99 --> nav away / cancel --> nav back showed the node count at 99 instead of cluster value
- In the cluster detail page `Config` tab, the select creds and Node Pool Taints (but not labels) components are editable / interactable
- ~I saw weirdness bringing up a cluster in UK South~
  - ~it might be something related to using 1.26, and/or multiple nodes~
  - ~i could bring up a single node 1.27 cluster in both East US and UK South~
  - ~in your testing did you try a zone with AV zone support other than East US, old kube version, multiple nodes, etc?~
  - ~do you know if zones have different config requirements?~

mantis-toboggan-md commented:
> I saw weirdness bringing up a cluster in UK South
> - it might be something related to using 1.26, and/or multiple nodes
> - i could bring up a single node 1.27 cluster in both East US and UK South
> - in your testing did you try a zone with AV zone support other than East US, old kube version, multiple nodes, etc?
> - do you know if zones have different config requirements?

What sort of weirdness did you see? I'd only tested US regions iirc but I've just tried to bring up a few in UK South and they all succeeded:
(all in zones 1 2 3, default vm size Standard_DS2_v2)

- 1 pool with 2 nodes, 26.10
- 2 pools with 1 node each, 26.10
- 1 pool with 1 node, 26.10
- 2 pools 2 nodes each, 1.27.7 <-- this one flashed an error faster than I could read it but ultimately provisioned

I doubled-checked both the azure docs and the process of creating an aks cluster through the azure portal directly and I didn't see any special requirements for availability zones.

mantis-toboggan-md commented:
>Tried creating a cluster, UK South, AV Zone 1, multiple roles assigned to a non-admin member. The cluster create succeeds but the 3 additional cluster role bindings fail with 404s

I've tried a bunch of times and I haven't been able to repro this, but I hit an issue adding cluster members as a standard user that I think is the same underlying problem. The error message in your screenshot is indicating that the namespace we expect to be created in the local cluster when a norman/mgmt cluster is created doesn't yet exist. In RKE2 provisioning, we wait for the management cluster to exist in the store before attempting to save CRTBs...we don't do that in this context because we're waiting for the norman cluster to save, which is the same underlying resource as the management cluster (right)? 

In the Ember world, we wait for the `InitialRolesPopulated` condition before applying after-save hooks...that seems to resolve my issue adding members as a standard user, and I think given that its the only condition we wait for in Ember, it's probably enough to fix the bug you saw as well...?

mantis-toboggan-md commented:
> Member Roles had a warning Note: Removing User will not remove their project permissions
> I think this is trying to say removing a cluster role will not mean users lose project/namespace roles (and therefor can still see cluster)?
> Update: Ok, lets track that in another issue. It's an important warning that we should probably be showing in other cases as well

My bad on this one, we do show this message in other cluster provisioning pages, the banner is just added outside the ClusterMembershipEditor component.

richard-cox reviewed (changes requested):
two new v small comments, conflicts aside everything else looks good!

richard-cox reviewed (approved):
My DO is poking me in the eye atm, so i couldn't run the latest changes.

The template change was unexpected, but i think makes sense. The others changes were trivial.

Mash that merge button!

mantis-toboggan-md commented:
> My DO is poking me in the eye atm, so i couldn't run the latest changes.
> 
> The template change was unexpected, but i think makes sense. The others changes were trivial.
> 
> Mash that merge button!

I should have left a comment about that one, sorry: the e2e tests caught a problem on the imported cluster page caused by my previous template change. But as you noted, it was a straightforward fix.
